src/gan/generator/lang/py/parse.py

```diff
@@ -0,0 +1,406 @@
+    elif p_except.match(ref_raw_code):
+        code = code.replace('try: pass\n', '').strip()
+    elif p_finally.match(ref_raw_code):
+        code = code.replace('try: pass\n', '').strip()
+
+    # remove ending pass
+    if code.endswith(':pass'):
+        code = code[:-len('pass')]
+
+    return code.strip()
+
+
+def add_root(tree):
+    root_node = ASTNode('root')
+    root_node.add_child(tree)
+
+    return root_node
+
+
+def parse(code):
+    """
+    parse a python code into a tree structure
+    code -> AST tree -> AST tree to internal tree structure
+    """
+
+    code = canonicalize_code(code)
+    py_ast = ast.parse(code)
+
+    tree = python_ast_to_parse_tree(py_ast.body[0])
+
+    tree = add_root(tree)
+
+    return tree
+
+
+def parse_raw(code):
+    py_ast = ast.parse(code)
+
+    tree = python_ast_to_parse_tree(py_ast.body[0])
+
+    tree = add_root(tree)
+
+    return tree
+
+
+def get_grammar(parse_trees):
+    rules = set()
+    # rule_num_dist = defaultdict(int)
+
+    for parse_tree in parse_trees:
+        parse_tree_rules, rule_parents = parse_tree.get_productions()
+        for rule in parse_tree_rules:
+            rules.add(rule)
+
+    rules = list(sorted(rules, key=lambda x: x.__repr__()))
+    grammar = PythonGrammar(rules)
+
+    logging.info('num. rules: %d', len(rules))
+
+    return grammar
+
+
+def tokenize_code(code):
+    token_stream = generate_tokens(StringIO(code).readline)
+    tokens = []
+    for toknum, tokval, (srow, scol), (erow, ecol), _ in token_stream:
+        if toknum == tk.ENDMARKER:
+            break
+        tokens.append(tokval)
+
+    return tokens
+
+
+def tokenize_code_adv(code, breakCamelStr=False):
+    token_stream = generate_tokens(StringIO(code).readline)
+    tokens = []
+    indent_level = 0
+    for toknum, tokval, (srow, scol), (erow, ecol), _ in token_stream:
+        if toknum == tk.ENDMARKER:
+            break
+
+        if toknum == tk.INDENT:
+            indent_level += 1
+            tokens.extend(['#INDENT#'] * indent_level)
+            continue
+        elif toknum == tk.DEDENT:
+            indent_level -= 1
+            tokens.extend(['#INDENT#'] * indent_level)
+            continue
+        elif len(tokens) > 0 and tokens[-1] == '\n' and tokval != '\n':
+            tokens.extend(['#INDENT#'] * indent_level)
+
+        if toknum == tk.STRING:
+            quote = tokval[0]
+            tokval = tokval[1:-1]
+            tokens.append(quote)
+
+        if breakCamelStr:
+            sub_tokens = re.sub(r'([a-z])([A-Z])', r'\1 #MERGE# \2', tokval).split(' ')
+            tokens.extend(sub_tokens)
+        else:
+            tokens.append(tokval)
+
+        if toknum == tk.STRING:
+            tokens.append(quote)
+
+    return tokens
+
+
+if __name__ == '__main__':
+    from nn.utils.generic_utils import init_logging
+    init_logging('misc.log')
+
+    # django_code_file = '/Users/yinpengcheng/Research/SemanticParsing/CodeGeneration/en-django/all.code'
```

Comment:
Everything from this point below... needed or not? The comments are one thing, but I'm concerned more with the `code = """` piece. Can we delete?